### PR TITLE
[JENKINS-64479] do not use JUL by default for logging

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-64479](https://issues.jenkins.io/browse/JENKINS-64479)
Logging during classloading is a surefire way to deadlock.
but we still want to have logs and in some cases we also want those logs
to be in the same place as other logs.

so introduce a new helper that will either output to System.out
(default) or use JUL (opt in) at the risk of deadlocks

### CODE REVIEW

**Review by aHenryJard — APPROVED** [review]
Looks good to me, as I saw the deadlock with my eyes so I think it's great if we can remove it :)

File: bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package org.jenkinsci.bytecode.helper;
+
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public final class LoggingHelper {
```
Comment (jtnord) [inline bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java:7]:
```suggestion
/**
 * Helper utility to direct logging to either System.out or the given Logger (if enabled).
 * We need this intermediate as logging can cause classloading and also involves locks, and the transformer is only ever
 * invoked during class loading - so we should not really use logging as we could end up in a deadlock (however there
 * are times where we really want to use it)
 */
public final class LoggingHelper {
```

**Review by jglick — COMMENTED** [review]
I guess? It is not obvious to me that this actually solves the issue; might just shift the trigger condition from one set of conditions to another.

Perhaps it would be better to retain `LoggingHelper` but have it always use JUL, but asynchronously:

* return early if log level too low
* else start a new `Thread` (avoid dep on executors etc.) to emit this one message later

The performance of actually logging an entry would not be great, but in production usage we expect this component to never log a single thing so I do not much care.

**Review by amuniz — APPROVED** [review]

**Comment by jtnord** [comment]
> I guess? It is not obvious to me that this actually solves the issue; might just shift the trigger condition from one set of conditions to another.

@jglick you are partly correct :-p

The only way to solve all possible deadlock is to not do anything here that could involve classloading from a different classloader.  As if we do there is nothing to say that the other classloader won;t be trying to load some class form the classloader that we are being callaed from.  So basically no `toString()` on anything that is not already loaded.

The actual observed deadlock was actually in a lock that logging held.  (the `LogHandler`) as its implementation is as follows...
https://github.com/openjdk/jdk/blob/00d80fdd862a01a78097fcf37e46a7254cc8c961/src/java.logging/share/classes/java/util/logging/StreamHandler.java#L193-L219

so it is a `synchronized` method and it is calling `toString()` on an object that in the observed case called something that needed a class loading whilst holding a lock on the LogHandler - so no other log recorders could be published, which blocks all logging calls where the logging should be published (because it is not async).

> Perhaps it would be better to retain `LoggingHelper` but have it always use JUL, but asynchronously:

so I thought about that and did not go down that route because often the interesting thing is a threadId (being able to correlate messages for a given thread is often vital when it comes to the BCT)

one possible way around that is to actually construct a `LogRecord` and to force all values to be populated (and immutable - which they should already be in the calls in this package) before punting off to logging, but it can still mean some interesting out of order log messages. its possible, but I went this route.  I can investigate this if you think it would be helpful..

> * return early if log level too low
> * else start a new `Thread` (avoid dep on executors etc.) to emit this one message later
> 
> The performance of actually logging an entry would not be great, but in production usage we expect this component to never log a single thing so I do not much care.

**Comment by jglick** [comment]
No strong opinion. The current patch should be fine so long as you are confident that it would actually prevent the deadlock, and that actual log messages from this component are rare enough in practice that use of stderr is not going to reduce diagnostic capacity in typical cases.

**Pull request closed without merging**